cmake/CMakeLists.txt

```diff
@@ -28,6 +28,11 @@ else()
   set(CMAKE_CXX_EXTENSIONS OFF)
 endif()
 
+# Quiet some noisy Intel warnings
+if (CMAKE_CXX_COMPILER_ID MATCHES Intel)
+  string(APPEND CMAKE_CXX_FLAGS " -diag-disable=2196")
```

Comment:
Could you add a comment here saying what kind of warning this disables?

Reply:
The Intel compiler has difficulty dealing with `noinline` member functions for template classes since it treats all template code as implicitly inlineable.  As an example:

```cxx
// foo.cxx
#include <iostream>

template<typename T>
class foo
{
public:
  void bar1(int arg) { std::cout << arg; }

  __attribute__((noinline)) void bar2(T arg) { std::cout << arg; }

  inline void bar3(T arg) { std::cout << arg; }
};

void baz()
{
  foo<double> f;

  f.bar1(42);    // OK
  f.bar2(42.0);  // Warning
  f.bar3(42.0);  // OK
}
```
```bash
$ icpc -c foo.cxx 
foo.cxx(21): warning #2196: routine is both "inline" and "noinline"
```

This is a pretty common pattern throughout protobuf and causes `warning #2196: routine is both "inline" and "noinline"` to spam the output with >800 warnings originating from over a quarter of the files in the codebase.  The issue itself seems to be more of a compiler problem in that icpc isn't able to produce noinline member functions of template classes so it seems most appropriate to just silence the warning.

Reply:
Added comment explaining the warning and why it's silenced.